modules/apps/frontend-js/frontend-js-components-sample-web/src/main/resources/META-INF/resources/js/SampleWalkthrough.js

```diff
@@ -18,34 +18,40 @@ import React from 'react';
 const WALKTHROUGH_CONFIG = {
 	closeOnClickOutside: false,
```

Comment:
Let me provide some context aboit this, just info. When we created this sample, we made the assumption that this piece of JSON was really part of a bigger, site-level JSON which would have a per-page structure like this:
```
[  "page1-url" : { // all the stuff we set in WALKTRHOUGH_CONFIG constant }, 
   "page2-url" : { // all the other stuff we'd like to show in this other page },
   ...
   "pagen-url" : { }
]
```
In this setting, when it comes to serve the walkthrough component properly configured, the dynamic include would grab the piece corresponding to current URL and use that portion to render the walktrhough.

With the multi-page setting, we need a different approach. Now, there's no just steps in a page. We have a site-level JSON containing one or more walkthroughs, then, each of it, may refer to one or more pages. 

So, current, I'm guessing that addition of `pages` entry serves that purpose as it's an array of page URLs with step identifiers right?

Reply:
Oh I see, yeah this is an idea so that the navigation of multi-pages per site can work, we need to know which steps are available on the other pages.

The initial logic of the implementation follows that the array of `steps` they must be declared sequentially so that we can move between the steps using indices, the calculation also to say the total number of steps and which step the user is in, the component is dependent on having visibility of all steps to calculate this, it could be done exposed via property as well.

I think to perform multi-page navigation with the structure you showed, we would need to have properties like `totalSteps` and maybe the `step x of x`. I don't think we could use indexes to navigate either, because when browsing we will need to reset the index to start from index zero of that page, we would also have to somehow make the index be used only on the page the user is in if he goes back to the previous page it is likely that some step will be rendered.

I think the main tradeoff here is that in this current implementation the `steps` property has all the website steps, which sends a lot of data over the network but simplifies implementation and is also simpler to manage. The other option is that we send less data, because we would only be sending the steps of the page but we would have more problems managing the step controls and knowing which step the user is in and etc... it is also very likely that it would be worse to organize the order of the steps.

I'm fine with which decision but I believe that this current one makes the general modal implementation simpler, the other one would have to have a lot more properties so that the runtime can render the information correctly. What do you think?

Reply:
The structure I shown is probably not the best to support multi-page walktrhughs, so we can safely dismiss it. Kill it with fire!

I think `pages` element is great to reflect the new concept. If I understood properly, we're indexing steps per page, so we can know what steps belong to what pages. Therefore, a walkthrough is then defined as a pair of elements: `pages` and `steps` plus some extra properties like `closeable` etc.

Let's ignore for a moment how much data we sent to the page: it doesn't matter if we include steps for a different page if that simplifies implementation. In the end we should expect a low number of steps right?

The thing is that all those steps refer to the _same_ walkthrough. So, in a site with multiple pages, we can define one walkthrough able to inform users about a great feature across several pages.

Now, imagine we need to show 2 separate features. For that case, we should add the concept of supporting more than 1 walktrhough per site. Meaning that we need an structure that replicates what we defined, something like:
```
{ 
   "walkthrough1" : {
       "pages": { },
       "steps": [ ]
   },
   "walkthrough2" : {
       "pages": { },
       "steps": [ ]
   }
}
```
Now, with this setting, we need to know which walkthrough we need to serve, because we may have that the same page includes steps for 2 separate walkthroughs. 

Few things to note:
- In the old, "mono-page" walktroughs, this was not ever an issue because we assumed that different pages could include separate walkthroughs if needed, and no single page could contain 2 different walktrhoughs
- We don't have to support this now, just wanted to note that defining more than one feature to showcase is something we'll be likely asked to support
- Supporting this looks a natural extension of the logic in this pull, maybe with the help of some cookies to make backend aware of what waktrhoughs have been consumed by the user.
- Then, we can send to the page the "current" walktrhough in place, along with all its steps. Back to the # of steps sent to the page, I don't think this is an issue.

Reply:
Yeah that makes sense to me too, I also think that the amount of steps being sent over the network is not a big issue, the amount will definitely be less than 100 and it will still be very little data I believe.

Having one or more different walkthroughs on the same site, can make sense too I believe that to cover this case it would be maybe even simple, instead of putting this on the component side in React, it could be the backend putting to render more than one component of the walkthrough.